text/0000-multisignature-outputs.md

```diff
@@ -0,0 +1,393 @@
+
+The derivation function:
+
+```
+id_hash = Blake2b("multisig_id" || SlateIDBytes || AmountBytesBE, 16)
+```
+
+where `SlateIDBytes` is the serialized byte representation of the Slate UUID, and `AmountBytesBE` is the 64-bit unsigned integer amount in 8-byte big-endian form.
+
+The multisig key ID is a BIP32-style identifier, constructed from the hash output (using python slice notation):
+
+```
+Identifier(depth: 4, id_hash[0:4], id_hash[4:8], id_hash[8:12], id_hash[12:16])
+```
+
+### Common nonce
+[common-nonce]: #common-nonce
+
+To build a multiparty bulletproof, a common nonce is required over the shared commitment to the transaction amount.
+
+The common nonce is derived using a Diffie-Hellman exchange of the public nonces of the participants.
+
+The result of the Diffie-Hellman exchange is hashed together with a domain separation string `multisig_common_nonce` using the SHA3-256 hashing algorithm:
+
+```
+// For party a:
+shared_key_in = Diffie-Hellman(sec_nonce_a, pub_nonce_b)
+common_nonce = SecretKey(Sha3_256("multisig_common_nonce" || shared_key_in_bytes))
+
+// For party b:
+shared_key_in = Diffie-Hellman(sec_nonce_b, pub_nonce_a)
+common_nonce = SecretKey(Sha3_256("multisig_common_nonce" || shared_key_in_bytes))
+```
+
+where `shared_key_in_bytes` is the byte representation of the serialized Secp256k1 secret key.
+
+SHA3-256 is used for its security, 32-byte output, and use in other parts of the Grin implementation (e.g. slatepack addresses).
+
+### Multiparty bulletproofs
+
+In this section, building multiparty bulletproofs is described. The process requires mutliple communication rounds to exchange the partial keys used for constructing the final proof.
+
+The protocol follows the specification originally proposed in the bulletproofs white paper by Buenz et al. [\[1\]](#references) (Ss. 4.5 "A Simple MPC Protocol for Bulletproofs"), and the implementation in the Grin fork of `secp256k1-zkp` [\[2\]](#references).
+
+#### init_send_tx
+
+In the first round, the initiator commits to value of the transaction using their `multisig_key`. The `multisig_key_id` is generated according to the previous section.
+
+```
+(multisig_key_a_pub, multisig_key_a_sec) = a.keychain().derive_key(multisig_key_id)
+
+part_commit_a = PedersenCommit(value, multisig_key_a_sec)
+```
+
+The partial commit is added to the slate, and sent to the receiver.
+
+#### receive_tx
+
+The receiver extracts the initiator's partial commit from the slate, and computes their own partial commit:
+
+```
+(multisig_key_b_pub, multisig_key_b_sec) = b.keychain().derive_key(multisig_key_id)
+
+part_commit_b = PedersenCommit(value, multisig_key_b_sec)
```

Comment:
Is it possible for B to cancel the pubkey of A here? e.g.
```
part_commit_b = PedersenCommit(value, multisig_key_b_sec) - (part_commit_a - value*H)
```
or something similar? The attempt would be to create a `commit_sum` output that contains only the blinding factor of the party B. I'm not able to follow the bulletproof part because I'm not familiar with this part, but perhaps the receiver could throw away the final bulletproof and simply construct the bulletproof alone since they know both values. This would also require B to also cancel the same key in the partial excess pubkey of B since A expects their private key to be a part of that.